src/comparison/balanced_pb.rs

```diff
@@ -0,0 +1,214 @@
+                    history_starting_index = segment_index;
+                }
+            } else {
+                history_starting_index = segment_index;
+            }
+        }
+    }
+
+    // TODO Reuse
+    let mut weighted_lists = Vec::new();
+    let mut overall_starting_index = -1;
+
+    for (current_index, (current_list, segment)) in
+        all_history.into_iter().zip(segments.iter()).enumerate()
+    {
+        let mut null_segment = false;
+        let current_pb_time = segment.personal_best_split_time()[method];
+        let previous_pb_time = if overall_starting_index >= 0 {
+            segments[overall_starting_index as usize].personal_best_split_time()[method]
+        } else {
+            None
+        };
+
+        // TODO More allocations :(
+        let mut final_list = current_list
+            .into_iter()
+            .filter(|&(index, _)| index == overall_starting_index)
+            .map(|(_, time)| time)
+            .collect::<Vec<_>>();
+
+        if !final_list.is_empty() {
+            overall_starting_index = current_index as i64;
+        } else if let Some(diff) = catch! { current_pb_time? - previous_pb_time? } {
+            final_list.push(diff);
+            overall_starting_index = current_index as i64;
+        } else {
+            null_segment = true;
+        }
+
+        if !null_segment {
+            let count = final_list.len();
+            // TODO More allocations FeelsBadMan at this point
+            let mut temp_list = final_list
+                .into_iter()
+                .enumerate()
+                .map(|(i, time)| (get_weight(i, count), time))
+                .collect::<Vec<_>>();
+
+            // TODO Aaarghhh >:(
+            let mut weighted_list = Vec::new();
+
+            if temp_list.len() > 1 {
+                temp_list
+                    .sort_unstable_by_key(|&(_, time)| OrderedFloat(time.total_milliseconds()));
+
+                let total_weight = temp_list.iter().map(|&(weight, _)| weight).sum::<f64>();
+                // TODO That's the smallest time's weight, not the actual
+                // smallest weight. Is that intended?
+                let smallest_weight = temp_list[0].0;
+                let range_weight = total_weight - smallest_weight;
+
+                let mut agg_weight = 0.0;
+                for (weight, time) in temp_list {
+                    agg_weight += weight;
+                    // TODO Possibly use extend instead for better reallocation behavior
+                    weighted_list.push((reweight(agg_weight, smallest_weight, range_weight), time));
+                }
+
+                // TODO What's the point of sorting this? temp_list was already
+                // sorted by the times and was inserted in that order.
+                // weighted_list was empty before.
+                weighted_list
+                    .sort_unstable_by_key(|&(_, time)| OrderedFloat(time.total_milliseconds()));
```

Comment:
@ROMaster2 this also seems very questionable. We sorted the temp_list before by the times and then store the elements in the weighted_list in the exact same order, so there doesn't seem to be a reason to sort them again.